File: Solutions/Problem108/Nag/Problem108SortedArrayToBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */

 const sortedArrayToBST = function(nums) {
    if (nums.length == 0) return null;
    
     const constructBinaryTree = (left, right) => {
    if(left > right) return null
    const midpoint = Math.floor(left + (right-left) /2);
    const node = new TreeNode (nums[midpoint], constructBinaryTree(left, midpoint-1, nums), constructBinaryTree(midpoint+1, right, nums));
    return node;
}
    return constructBinaryTree(0, nums.length-1);
};

File: Solutions/Problem108/Nag/Problem108SortedArrayToBST.java
//class TreeNode {
//    int val;
//    TreeNode left;
//    TreeNode right;
//
//    TreeNode() {
//    }
//
//    TreeNode(int val) {
//        this.val = val;
//    }
//
//    TreeNode(int val, TreeNode left, TreeNode right) {
//        this.val = val;
//        this.left = left;
//        this.right = right;
//    }
//
//}

public class Problem108SortedArrayToBST {


    public static TreeNode sortedArrayToBST (int[] nums){
        if(nums.length == 0) return null;
        return Problem108SortedArrayToBST.constructBST(nums, 0, nums.length -1);
    }

    public static TreeNode constructBST(int[]nums, int left, int right){
        // base case of when to stop constructing tree
        if(left > right) return null;
        int midpoint = left + (right-left)/2;
        TreeNode node =  new TreeNode(nums[midpoint]);
        node.left = constructBST(nums, left, midpoint-1);
        node.right = constructBST(nums, midpoint+1, nums.length-1);
        return node;
    }

}
